### PR TITLE
[FLINK-23218][coordination] Distribute the ShuffleDescriptors via blob server

### PR DESCRIPTION
**Pull request by Thesharing**
## What is the purpose of the change

*This pull request is the second part of optimizing the deployment of tasks. In this pull request, we propose to distribute the cached ShuffleDescriptors via the blob server if their overall size has exceeded the threshold. For more details please check FLINK-23218. The first part of the optimization is located at #16314.*

*In this pull request we remove the cache from blob server on JobManager once it's no longer required. To remove the cache from the blob cache located on TaskManager, please check #16498.*

## Brief change log

  - *The cached ShuffleDescriptors are stored in the blob server if its size exceeded the threshold.*
  - *The cached ShuffleDescriptors will be removed from the blob server once it's no longer used on JobManager.*
  - *TaskManagers can retrieve the ShuffleDescriptors from the blob cache.*

## Verifying this change

  - *Added unit tests to test the ShuffleDescriptors are cached and removed correctly.*
  - *Added unit tests to verify that deleting files on blob server works as expected.*
  - *Manually verified the change by running an end-to-end test. The job has two JobVertices. One is source, while the other is sink. The parallelism is 8k and 16k. They are connected with an all-to-all edge. Both streaming mode and batch mode are tested.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (**yes** / no / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/InternalExecutionGraphAccessor.java [hunk 1]
```diff
@@ -103,4 +104,6 @@ void notifySchedulerNgAboutInternalTaskFailure(
     ExecutionVertex getExecutionVertexOrThrow(ExecutionVertexID id);
 
     IntermediateResultPartition getResultPartitionOrThrow(final IntermediateResultPartitionID id);
+
+    void notifyBlobWriterDeleteCache(List<PermanentBlobKey> blobKeys);
```
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/InternalExecutionGraphAccessor.java:108]: Maybe name it as `deleteBlobCache(...)`?

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobServer.java [hunk 2]
```diff
@@ -778,6 +778,50 @@ boolean deleteInternal(@Nullable JobID jobId, TransientBlobKey key) {
...
+    private boolean deleteInternal(JobID jobId, PermanentBlobKey key) {
+        final File localFile =
+                new File(
+                        BlobUtils.getStorageLocationPath(storageDir.getAbsolutePath(), jobId, key));
+
+        readWriteLock.writeLock().lock();
+
+        try {
+            boolean deleteLocally = true;
+            if (!localFile.delete() && localFile.exists()) {
```
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobServer.java:798]: Will an exception be thrown if the file was already deleted by another thread?
Should we check the existence of the file before try deleting it?

File: flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobServerDeleteTest.java [hunk 3]
```diff
@@ -251,6 +251,104 @@ private void testDeleteTransientFails(@Nullable final JobID jobId) throws IOExce
         }
     }
 
+    @Test
+    public void testDeletePermanent() throws IOException {
```
Comment (zhuzhurk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobServerDeleteTest.java:255]: Is it possible to reuse the code of `testDeleteTransient`, `testDeleteTransientAlreadyDeleted` and `testDeleteTransientFails`?

File: flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobWriterTest.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+import java.io.IOException;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+/** Tests for {@link BlobWriter}. */
+public class BlobWriterTest extends TestLogger {
+
+    @Test
+    public void testSerializeAndTryOffload() throws IOException, ClassNotFoundException {
```
Comment (zhuzhurk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobWriterTest.java:38]: I think it does not make much sense to add a test case which verifies a test class. While the test class is only used by its own unit test.
I think the newly added tests in `BlobServerDeleteTest` are enough.

**Review by zhuzhurk — COMMENTED** [review]
Thanks for opening this PR! @Thesharing 
I have a few comments. Please take a look.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/InternalExecutionGraphAccessor.java (hunk 1 above)
Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/InternalExecutionGraphAccessor.java]: Or maybe `deleteBlobs(...)`?

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobServer.java (hunk 2 above)
Comment (Thesharing) [reply flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobServer.java:798]: No exceptions will be thrown if the delete failed. 

`java.io.File#delete` won't thrown any thing if the file cannot be deleted. On the contrary, `java.nio.file.Files#delete` will throw `IOException` if the file cannot be deleted. In fact `deleteInternal(JobID jobId, PermanentBlobKey key)` is somehow the same as `deleteInternal(@Nullable JobID jobId, TransientBlobKey key)`.

**Review by zhuzhurk — APPROVED** [review]
The change ac8aae3283bc8d17e47b71a0a68052d96ea1a77c looks good to me.
Will merge it once #16314 is ready.

**Comment by zhuzhurk** [comment]
Rebased onto latest master and fixed checkstyle violations (see https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=20982&view=logs&j=52b61abe-a3cc-5bde-cc35-1bbe89bb7df5&t=54421a62-0c80-5aad-3319-094ff69180bb).

**Comment by Thesharing** [comment]
> Rebased onto latest master and fixed checkstyle violations (see https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=20982&view=logs&j=52b61abe-a3cc-5bde-cc35-1bbe89bb7df5&t=54421a62-0c80-5aad-3319-094ff69180bb).

Sorry for being careless. My checkstyle plugin in IDEA seems to be broken.

**Comment by Thesharing** [comment]
@flinkbot run azure

**Pull request closed without merging**